refactor configuration

joshgav opened the pull request:
This updates the configuration system used in this repo. The `helpers` and `iam` dirs are removed in favor of new `internal` packages for configuration.

The `resources` and `network` packages are updated already to use the new config system and their tests (`go test -v ./resources/` and `go test -v ./network/`) pass. I'll continue updating the other dirs but the core is ready for feedback now.

cc @marstr

joshgav commented:
fixes #194

marstr commented on internal/config/config.go:
```diff
@@ -0,0 +1,116 @@
+
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/marstr/randname"
```
This makes me very happy 😃

marstr commented on internal/iam/authorizers.go:
```diff
@@ -0,0 +1,210 @@
+		return keyvaultAuthorizer, nil
+	}
+
+	// BUG: default value for KeyVaultEndpoint is wrong
```
Can you find/create an issue to track fixing this?

joshgav replied:
Yes, absolutely plan to, would like to wait till I can test and file a descriptive issue, hence marked as `BUG` for now.

marstr commented on internal/iam/authorizers.go:
```diff
@@ -0,0 +1,210 @@
+		deviceConfig.AADEndpoint = alternateEndpoint.String()
+		a, err = deviceConfig.Authorizer()
+	default:
+		log.Fatalln("invalid token type specified")
```
I'm never a fan of `os.Exit(1)`, like is used by `log.Fatalln`. It prevents any deferred actions from executing all the way up the stack.

I would rather see this clause as:

``` Go
return nil, errors.New("invalid token type specified")
```

or even better, create a global variable `var ErrInvalidTokenType = errors.New("invalid token type specified")`  then return that. :)

marstr commented on network/hybrid/network.go:
```diff
@@ -242,37 +244,37 @@ func CreateNetworkInterface(cntx context.Context, netInterfaceName, nsgName, vne
-func GetNetworkSecurityGroup(cntx context.Context, nsgName string) (network.SecurityGroup, error) {
-	environment, _ := azure.EnvironmentFromURL(helpers.ArmEndpoint())
+func GetNetworkSecurityGroup(ctx context.Context, nsgName string) (network.SecurityGroup, error) {
+	environment, _ := azure.EnvironmentFromURL(config.Environment().ResourceManagerEndpoint)
```
Unless there's something I don't see, this should be simplified to:

``` Go
 environment := config.Environment()
```

(Same for all of the other instances in this file)

marstr commented on network/hybrid/network_test.go:
```diff
@@ -7,14 +7,17 @@ package network
 	hybridresources "github.com/Azure-Samples/azure-sdk-for-go-samples/resources/hybrid"
+
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/internal/config"
+	internalutil "github.com/Azure-Samples/azure-sdk-for-go-samples/internal/util"
```
[nit] I'd rather this just be `util`

joshgav replied:
My thought was to use the name `internalutil` consistently across the samples, expecting other local `util` packages to sometimes be used. I changed per your preference though, cause I think the practice we're going for is to always use the actual import name unless impossible.

marstr commented on internal/config/config.go:
```diff
@@ -0,0 +1,116 @@
+	if len(affix) > 0 {
+		affix += "-"
+	}
+	return randname.GenerateWithPrefix(fmt.Sprintf("%s-%s", BaseGroupName(), affix), 5)
```
[nit] I'd make affix essentially an optional parameter by doing this:

``` Go
func GenerateGroupName(affix ...string) string {
    builder := bytes.NewBufferString(BaseGroupName())
    builder.WriteRune('-')
    for _, clause := range affix {
        fmt.Fprint(builder, clause)
        builder.WriteRune('-')
    }

    return randname.GenerateWithPrefix(builder.String(), 5)
}
```

marstr commented on internal/config/env.go:
```diff
@@ -0,0 +1,50 @@
+	// these must be provided by environment
+	var err1, err2, err3, err4 error
+	clientID, err1 = envy.MustGet("AZURE_CLIENT_ID")
+	clientSecret, err2 = envy.MustGet("AZURE_CLIENT_SECRET")
```
I'm surprised these are required. I thought there was some ability to use Device flow?

joshgav replied:
Tried to get fancy with error handling but will change to handle each and exclude client secret when device flow is used.

I see now why a framework for configuration is helpful, it's hard to handle precedence and relationships between different config sources smoothly.

marstr commented on internal/config/env.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	// we can choose defaults for these
+	groupName = envy.Get("AZURE_GROUP_NAME", "azure-go-samples")
```
I'd make this random to reduce the chance that test runs are influenced by resources that may already exist.

``` Go
groupName = envy.Get("AZURE_GROUP_NAME", randname.GenerateWithPrefix("azure-go-samples-", 6)
```

joshgav replied:
I was thinking of addressing the collision issue *and* the use of a global var for group name by deprecating `AZURE_GROUP_NAME` and using `AZURE_BASE_GROUP_NAME` instead, see next line. Then we can use `GenerateGroupName(affixes ...string)` in packages to get a derived name for local use.

I noted the deprecation in a comment on the accessor and will add a comment here too.

marstr commented on internal/config/env.go:
```diff
@@ -0,0 +1,50 @@
+
+	// we can choose defaults for these
+	groupName = envy.Get("AZURE_GROUP_NAME", "azure-go-samples")
+	baseGroupName = envy.Get("AZURE_BASE_GROUP_NAME", groupName)
```
[nit] I'd make a const for each of the environment variables which can control our test runs. I'd put all of those constants in a single `const` block for the sake of documentation.

joshgav replied:
I think that's what I this is: https://github.com/joshgav/azure-sdk-for-go-samples/blob/33ba00a4453a05b4c9bf84b4f2042a04a1c47708/internal/config/config.go#L14-L30

It's a `var` block cause the values have to be manipulated by `env.go` and `flags.go`.

marstr commented on network/hybrid/network_test.go:
```diff
@@ -92,13 +119,18 @@ func ExampleCreatePublicIP() {
 	_, err := hybridresources.CreateGroup(ctx)
 	if err != nil {
-		helpers.PrintAndLog(err.Error())
+		internalutil.PrintAndLog(err.Error())
```
I'd like to get rid of `PrintAndLog`. We could use `logrus` or something similar instead.

marstr replied:
Doing this would help give people guidance of real libraries they could use for logging in their projects, instead of just showing them something we wrote in a cage.

joshgav replied:
I'd like to change the examples to test cases and use `testing.T` and have started to convert some already in this PR to demonstrate. Do you think that's a good idea?

joshgav commented:
Thanks @marstr! Addressed your comments. Will now proceed to update all packages so tests pass.

I'm thinking of replacing the `ExampleXxx()` tests with `TestXxx(t *testing.T)` tests, see for example https://github.com/joshgav/azure-sdk-for-go-samples/blob/b430acf30379bff21dbd133ae3aa1ad2b0a60111/resources/resources_test.go#L47. There's a bit more structure with `testing.T` which would come in handy, such as replacing `PrintAndLog` :). What do you think?

joshgav commented:
Tests are underway here: https://travis-ci.org/Azure-Samples/azure-sdk-for-go-samples/jobs/407320872. I believe the ones that fail are auth and permissions related and I'll troubleshoot. Otherwise I think this work is ready to merge. @vladbarosan @jhendrixMSFT @marstr if you'd like can you please review? Thanks!

joshgav commented:
fixes #207

vladbarosan commented:
> 

@ can you open an issue to track making the test run daily ? maybe hook it up with the vsts sample config

---
Refers to: .dockerignore:3 in 4e95ecd. [](commit_id = 4e95ecd68b1c0969dc8e7b6bf3a3b2f7e5ecdc76, deletion_comment = True)

vladbarosan commented on compute/compute_test.go:
```diff
@@ -8,158 +8,121 @@ package compute
 	"log"
 	"os"
+	"strings"
 	"testing"
```
>esting" [](start = 3, length = 7)

maybe move back to separate folders

joshgav replied:
oops, forgot to do this. will do it in next update.

joshgav replied:
tracked in #210

vladbarosan commented:
@joshgav approved so you can merge when ready, but please create issues for the work remaining ( e.g. the authorizers, running the samples etc )

joshgav commented:
Opened #208 and #209 to track related tasks, will merge this now. Thanks!
